File: aoc-py/solutions/day18.py
"""
Day 18: RAM Run

https://adventofcode.com/2024/day/18
"""
__all__ = ('Day18',)

from typing import ClassVar
from collections import deque

from ..solution import Solution
from ..utils import neighbors_4

class Day18(Solution):
    NAME: ClassVar[str] = 'RAM Run'

    def _parse_corrupted(self, inp: str) -> list[tuple[int, ...]]:
        return [
            tuple(int(x) for x in line.split(','))
            for line in inp.splitlines()
        ]

    def _find_path(self, corrupted: list[tuple[int, ...]], bytes_fell: int) -> int | None:
        n_rows = 71
        n_cols = 71

        corrupted_set = set(corrupted[:bytes_fell])

        to_check = deque([(0, 0, 0)])
        seen = {(0, 0)}

        while to_check:
            row, col, steps = to_check.popleft()

            if row == n_rows - 1 and col == n_cols - 1:
                return steps

            for next_coord in neighbors_4(row, col):
                next_row, next_col = next_coord

                if (
                    next_row in range(n_rows)
                    and next_col in range(n_cols)
                    and next_coord not in corrupted_set
                    and next_coord not in seen
                ):
                    to_check.append((*next_coord, steps + 1))
                    seen.add(next_coord)

    def part_one(self, inp: str) -> int:
        corrupted = self._parse_corrupted(inp)

        assert (steps := self._find_path(corrupted, 1024)) is not None
        return steps

    def part_two(self, inp: str) -> str:
        corrupted = self._parse_corrupted(inp)

        # Lower bound: Part 1 guarantees that the first 1024 bytes will not block the exit
        left = 1024
        # Upper bound: all bytes fell
        right = len(corrupted) - 1

        # binary search only having `mid` number of blocks fall
        #
        # As more bytes fall sequentially -> more likely the path is to be blocked (hence our searching medium is "sorted")
        while left <= right:
            mid = (left + right) // 2

            if self._find_path(corrupted, mid) is None:
                # we blocked the exit! Try even less corrupted bytes and see if the exit is still blocked
                # (optimize as we need to find the first byte that does this)
                right = mid - 1
            else:
                # exit still reachable: Try having more corrupted bytes fall in
                left = mid + 1

        # The first byte that when fell, will block the exit
        byte = corrupted[mid]
        return '{},{}'.format(*byte)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 302
        assert p2 == '24,32'

File: src/bin/day18.rs
//! Day 18: RAM Run
//!
//! <https://adventofcode.com/2024/day/18>

use std::{
    collections::{HashSet, VecDeque},
    fmt::Display
};
use aoc_2024::{Solution, neighbors_4};

pub struct Day18;

impl Day18 {
    fn parse_corrupted<T: Display>(inp: T) -> Vec<(usize, usize)> {
        inp
            .to_string()
            .lines()
            .map(|line| {
                let (x, y) = line.split_once(',').unwrap();
                (
                    x.parse::<usize>().unwrap(),
                    y.parse::<usize>().unwrap(),
                )
            })
            .collect()
    }

    fn find_path(corrupted: &Vec<(usize, usize)>, bytes_fell: usize) -> Option<usize> {
        let rows = 0..=70;
        let cols = 0..=70;

        let corrupted_set = HashSet::<&(usize, usize)>::from_iter(&corrupted[..bytes_fell]);

        let mut to_check = VecDeque::new();
        to_check.push_back((0, 0, 0));

        let mut seen = HashSet::new();
        seen.insert((0, 0));

        while let Some((row, col, steps)) = to_check.pop_front() {
            if row == *rows.end() && col == *cols.end() {
                return Some(steps);
            }

            for next_coord @ (next_row, next_col) in neighbors_4(row, col) {
                if rows.contains(&next_row)
                    && cols.contains(&next_col)
                    && !corrupted_set.contains(&next_coord)
                    && seen.insert(next_coord)
                {
                    to_check.push_back((next_row, next_col, steps + 1));
                }
            }
        }
        None
    }
}

impl Solution for Day18 {
    const NAME: &'static str = "RAM Run";

    type OutputP2 = String;

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        Self::find_path(
            &Self::parse_corrupted(inp),
            1024,
        )
        .unwrap()
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        let corrupted = Self::parse_corrupted(inp);

        let mut left = 1024;
        let mut right = corrupted.len() - 1;
        let mut mid = 0;

        while left <= right {
            mid = (left + right) / 2;

            if Self::find_path(&corrupted, mid).is_none() {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        let byte = corrupted[mid];
        return format!("{},{}", byte.0, byte.1);
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 302);
        assert_eq!(p2, "24,32");
    }
}

fn main() {
    aoc_2024::run_day(18, &Day18);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}